tikv/gc.go

```diff
@@ -94,72 +95,115 @@ func (s *KVStore) resolveLocks(ctx context.Context, safePoint uint64, concurrenc
+	// otherwise create it to prevent panic.
+	if l.bo == nil {
+		l.bo = NewGcResolveLockMaxBackoffer(ctx)
+	}
+	return batchResolveLocksInARegion(l.bo, l.GetStore(), locks, loc)
+}
+
+func (l *BaseLockResolver) ScanLocks(ctx context.Context, key []byte, maxVersion uint64) ([]*txnlock.Lock, *locate.KeyLocation, error) {
+	// every time ScanLocks need a new backoffer
+	l.bo = NewGcResolveLockMaxBackoffer(ctx)
+	return scanLocksInRegionWithStartKey(l.bo, l.GetStore(), key, maxVersion, GCScanLockLimit)
+}
+
+func (l *BaseLockResolver) GetStore() Storage {
+	return l.store
+}
+
+// GCLockResolver is used for GCWorker and log backup advancer to resolve locks.
+type GCLockResolver interface {
+	// ResolveLocks tries to resolve expired locks.
+	// 1. For GCWorker it will scan locks for all regions before *safepoint*,
+	// and force remove locks. rollback the txn, no matter the lock is expired of not.
+	// 2. For log backup advancer, it will scan all locks for a small range.
+	// and it will check status of the txn. resolve the locks if txn is expired, Or do nothing.
+	ResolveLocks(context.Context, []*txnlock.Lock, *locate.KeyLocation) (*locate.KeyLocation, error)
+
+	// ScanLocks return locks and location with given start key
+	ScanLocks(context.Context, []byte, uint64) ([]*txnlock.Lock, *locate.KeyLocation, error)
+
+	// GetStore is used to get store to GetRegionCache and SendReq for this lock resolver.
+	GetStore() Storage
+}
+
 // We don't want gc to sweep out the cached info belong to other processes, like coprocessor.
-const gcScanLockLimit = txnlock.ResolvedCacheSize / 2
+const GCScanLockLimit = txnlock.ResolvedCacheSize / 2
 
-func (s *KVStore) resolveLocksForRange(ctx context.Context, safePoint uint64, startKey []byte, endKey []byte) (rangetask.TaskStat, error) {
+func ResolveLocksForRange(ctx context.Context, uuid string, resolver GCLockResolver, maxVersion uint64, startKey []byte, endKey []byte) (rangetask.TaskStat, error) {
 	// for scan lock request, we must return all locks even if they are generated
 	// by the same transaction. because gc worker need to make sure all locks have been
 	// cleaned.
-
 	var stat rangetask.TaskStat
 	key := startKey
-	bo := NewGcResolveLockMaxBackoffer(ctx)
 	for {
 		select {
 		case <-ctx.Done():
 			return stat, errors.New("[gc worker] gc job canceled")
 		default:
 		}
-
-		locks, loc, err := s.scanLocksInRegionWithStartKey(bo, key, safePoint, gcScanLockLimit)
+		locks, loc, err := resolver.ScanLocks(ctx, key, maxVersion)
 		if err != nil {
 			return stat, err
 		}
 
-		resolvedLocation, err1 := s.batchResolveLocksInARegion(bo, locks, loc)
-		if err1 != nil {
-			return stat, err1
+		resolvedLocation, err := resolver.ResolveLocks(ctx, locks, loc)
+		if err != nil {
+			return stat, err
 		}
 		// resolve locks failed since the locks are not in one region anymore, need retry.
 		if resolvedLocation == nil {
 			continue
 		}
-		if len(locks) < gcScanLockLimit {
+		if len(locks) < GCScanLockLimit {
```

Comment:
If this function is going to be used by TiDB, please also see if TiDB has its own settings of the scan limit. If so, consider making it customizable.

Reply:
TiDB has a `mergeLockScanner` which use `gcScanLockLimit` to scan locks.

Reply:
They both use the same setting 128. I think we can unified here